mock/ecs_client.go

```diff
@@ -2,67 +2,673 @@ package mock
+	var exported []*ecs.KeyValuePair
+	for k, v := range envVars {
+		exported = append(exported, &ecs.KeyValuePair{
+			Name:  utility.ToStringPtr(k),
+			Value: utility.ToStringPtr(v),
+		})
+	}
+	return exported
+}
+
+func newSecrets(secrets []*ecs.Secret) map[string]string {
+	converted := map[string]string{}
+	for _, secret := range secrets {
+		if secret == nil {
+			continue
+		}
+		converted[utility.FromStringPtr(secret.Name)] = utility.FromStringPtr(secret.ValueFrom)
+	}
+	return converted
+}
+
+func exportSecrets(secrets map[string]string) []*ecs.Secret {
+	var exported []*ecs.Secret
+	for k, v := range secrets {
+		exported = append(exported, &ecs.Secret{
+			Name:      utility.ToStringPtr(k),
+			ValueFrom: utility.ToStringPtr(v),
+		})
+	}
+	return exported
+}
+
+// ECSService is a global implementation of ECS that provides a simplified
+// in-memory implementation of the service that only stores metadata and does
+// not orchestrate real containers or container instances. This can be used
+// indirectly with the ECSClient to access or modify ECS resources, or used
+// directly.
+type ECSService struct {
+	Clusters map[string]ECSCluster
+	TaskDefs map[string][]ECSTaskDefinition
+}
+
+// GlobalECSService represents the global fake ECS service state.
+var GlobalECSService ECSService
+
+func init() {
+	GlobalECSService = ECSService{
+		Clusters: map[string]ECSCluster{},
+		TaskDefs: map[string][]ECSTaskDefinition{},
+	}
+}
+
 // ECSClient provides a mock implementation of a cocoa.ECSClient. This makes
 // it possible to introspect on inputs to the client and control the client's
 // output. It provides some default implementations where possible.
-type ECSClient struct{}
+type ECSClient struct {
+	RegisterTaskDefinitionInput  *ecs.RegisterTaskDefinitionInput
+	RegisterTaskDefinitionOutput *ecs.RegisterTaskDefinitionOutput
+
+	DeregisterTaskDefinitionInput  *ecs.DeregisterTaskDefinitionInput
+	DeregisterTaskDefinitionOutput *ecs.DeregisterTaskDefinitionOutput
+
+	ListTaskDefinitionsInput  *ecs.ListTaskDefinitionsInput
+	ListTaskDefinitionsOutput *ecs.ListTaskDefinitionsOutput
+
+	RunTaskInput  *ecs.RunTaskInput
+	RunTaskOutput *ecs.RunTaskOutput
+
+	DescribeTasksInput  *ecs.DescribeTasksInput
+	DescribeTasksOutput *ecs.DescribeTasksOutput
+
+	ListTasksInput  *ecs.ListTasksInput
+	ListTasksOutput *ecs.ListTasksOutput
+
+	StopTaskInput  *ecs.StopTaskInput
+	StopTaskOutput *ecs.StopTaskOutput
+
+	CloseError error
+}
 
 // RegisterTaskDefinition saves the input and returns a new mock task
 // definition. The mock output can be customized. By default, it will create a
 // cached task definition based on the input.
 func (c *ECSClient) RegisterTaskDefinition(ctx context.Context, in *ecs.RegisterTaskDefinitionInput) (*ecs.RegisterTaskDefinitionOutput, error) {
-	return nil, errors.New("TODO: implement")
+	c.RegisterTaskDefinitionInput = in
+
+	if c.RegisterTaskDefinitionOutput != nil {
```

Comment:
This seems a little weird, we always overwrite the input but never the output. Shouldn't we just overwrite both?

Reply:
This is because the ECS client is a combination of a testing spy and a testing stub (as defined by [Martin Fowler](https://martinfowler.com/articles/mocksArentStubs.html#TheDifferenceBetweenMocksAndStubs)).

This "mock" happens to be a mishmash of a spy (which records inputs), a stub (which provides canned answers to requests), and a fake testing implementation (backed by the in-memory ECS service).